### PR TITLE
Improve FlexSearch indexing

### LINKED ISSUE #31: Search index is broken for multilingual sites
## Description

The current code always uses the same search index expected to live at `/search-index.json` regardless of page language:

https://github.com/gethyas/doks-core/blob/8d92d1a8df87610adde025235d12b37f48651358/assets/js/flexsearch.js#L110

This path is even wrong when `defaultContentLanguageInSubdir = true` in `hugo.toml`.

## Expected result

`flexsearch.js` should use the search index of the current page language.

## Actual result

See description above.

## Suggested solution

We could change

https://github.com/gethyas/doks-core/blob/8d92d1a8df87610adde025235d12b37f48651358/assets/js/flexsearch.js#L110

to

```js
    var pathName = window.location.pathname;
    var pathLang = pathName.substring(0, pathName.indexOf("/", 1)+1);
    fetch(pathLang.concat("search-index.json"))
```

but this only works if `defaultContentLanguageInSubdir = true`.

To have a solution that also works with `defaultContentLanguageInSubdir = false`, I guess we'd need to pre-process the `flexsearch.js` asset file with Hugo leveraging [`resources.ExecuteAsTemplate`](https://gohugo.io/hugo-pipes/resource-from-template/) to take the value of `defaultContentLanguageInSubdir` into account.

What do you think?

### PR DESCRIPTION
**Pull request by salim-b**
## Summary

Improves the JS config file as well as the JSON data structure from which FlexSearch's document index is built:

- `assets/js/flexsearch.js` is now a Hugo template which gets rendered using [`ExecuteAsTemplate`](https://gohugo.io/hugo-pipes/resource-from-template/) (I've added a dedicated partial `esbuild-template.html` for future usage). This allows more powerful search configuration like the new `indexSummary` option. Also fixes #31.

- Tags are now structured data (array). I have a feeling that this could improve search performance a little bit (depending on the number of tags), but I didn't actually run any benchmarks. I verified that there are no regressions regarding search results (results stayed the same).

- Default date format is now `"2006-01-02"` ([ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)), which I think makes more sense since a) one *usually* doesn't search for content creation dates and b) if one still does, they will likely get better results (less false positives) when using ISO 8601 dates instead of a combination of month names and numeric days.

- Set indexing options [*per field*](https://github.com/nextapps-de/flexsearch#the-document-descriptor) to tweak options for date field.

- Optional custom date format string is now defined in `site.Params.searchDateFormat` instead of `home.Params.dateformat`. This seems like a better place (and more intuitive variable name) for users. It now also supports [date format layout identifiers](https://gohugo.io/functions/dateformat/#datetime-formatting-layouts) like `:date_full` which get properly localized.

## To do

Optional custom date format string should be documented for users. It can either be set site-wide via `searchDateFormat` in `config/_default/params.toml` or per-language via `<LanguagePrefix>.params.searchDateFormat` in `config/_default/languages.toml`.

## Checks

- [x] Read [Create a Pull Request](https://gethyas.com/docs/contributing/how-to-contribute/#create-a-pull-request)
- [x] Supports all screen sizes (if relevant)
- [x] Supports both light and dark mode (if relevant)
- [ ] Passes `npm run test`

### CODE REVIEW

**Comment by h-enk** [comment]
Nice — I'll run some tests with your settings later

Related to this and following up on  https://github.com/gethyas/create-hyas/pull/14 I did some tests and found a major improvement by not indexing the content, but the summary (which is derived from the content, the length can be set in `hugo.toml`).

```txt
# pages = 2000
summarylength = 20 # 70 (default)

current settings
---------------------
index content?			true
index summary?			false
search-index.json size (KB)	5,728

proposed settings
---------------------
index content?			false
index summary?			true
search-index.json size (KB)	601
```

Changes:
- Replace "content" with "summary": https://github.com/gethyas/doks-core/blob/main/assets/js/flexsearch.js#L25
- Remove "content: https://github.com/gethyas/doks-core/blob/main/layouts/home.searchindex.json#L8

These settings would be ideal for large sites. If you have a smaller site, you might want to index the content i.s.o. the summary.

It would be nice to make this an option, I guess. What do you think?

**Comment by salim-b** [comment]
> It would be nice to make this an option, I guess. What do you think?

I think it's not desirable for most Doks use cases to only index the first n words of content... but why not give people the freedom to choose themselves? 🗽

See the updated PR.

**Comment by h-enk** [comment]
Great piece of work! I'll have a closer look later.

File: layouts/partials/footer/script-footer.html
Comment (h-enk) [inline layouts/partials/footer/script-footer.html:1]:
```html
{{ $searchLimit := site.Params.doks.searchLimit -}}
```

File: assets/js/flexsearch.js
Comment (h-enk) [inline assets/js/flexsearch.js:1]:
```js
        {
          field: {{ if site.Params.doks.indexSummary }}'summary'{{ else }}'content'{{ end }}
        },
```

But, not (yet) a param?

Comment (salim-b) [reply assets/js/flexsearch.js:1]: Fixed now :)

File: layouts/partials/footer/script-footer.html
Comment (salim-b) [reply layouts/partials/footer/script-footer.html:1]: ?

Comment (h-enk) [reply layouts/partials/footer/script-footer.html:1]: Sorry about that/forget it — I thought the param was not in place, but it's there: https://github.com/gethyas/create-hyas/blob/main/template-doks/config/_default/params.toml#L25

**Pull request merged**